### DISCUSSION TITLE
Merge Meshes

### DISCUSSION THREAD

**Discussion by reox** [nschloe/meshio#1422] [Q&A]
I just wanted to use meshio to write me nice vtk files for some meshes, and therefore I need to merge multiple meshes together.
I did not found a direct way to do that, thus I wrote this (very simplified and not generally usable) function:
```
import meshio

def merge_meshes(a: meshio.Mesh = None, b: meshio.Mesh = None) -> meshio.Mesh:
    if a is None:
        return b
    if b is None:
        return a
    n_vert = a.points.shape[0]
    vertices = np.vstack((a.points, b.points))
    all_cell_types = a.cells_dict.keys() & b.cells_dict.keys()
    cells = dict()
    for cell_type in all_cell_types:
        if cell_type in a.cells_dict and cell_type in b.cells_dict:
            cells[cell_type] = np.vstack((a.cells_dict[cell_type], b.cells_dict[cell_type] + n_vert))
        elif cell_type in a.cells_dict:
            cells[cell_type] = a.cells_dict[cell_type]
        elif cell_type in b.cells_dict:
            cells[cell_type] = b.cells_dict[cell_type] + n_vert
    return meshio.Mesh(vertices, cells)
```

It simply concatenates the point vectors and adjusts the numbering in the second mesh.
Because that could be useful for other things as well, I wondered, if I overlooked something and there is more generic way already implemented?

**Comment by adtzlr** [comment]
Hi, such a function would indeed be nice - however, it is probably out of scope. For my understanding, meshio is just for import/export - not mesh manipulation.

Personally, I'm using the `stack()`-method of a [`MeshContainer`](https://felupe.readthedocs.io/en/latest/felupe/mesh.html#felupe.MeshContainer) from [adtzlr/felupe](https://github.com/adtzlr/felupe) - with partial support from meshio-meshes. Note that a FElupe mesh supports only one cell-block.

```python
import meshio
import felupe as fem

# container = fem.mesh.read("filename.vtk")  # read a mesh using meshio

meshes = [fem.Rectangle(n=2), fem.Rectangle(n=2)]
container = fem.MeshContainer(meshes, merge=False)  # optional merge of duplicated points
mesh = container.stack()
```

```ipython
>>> mesh.points
array([[0., 0.],
       [1., 0.],
       [0., 1.],
       [1., 1.],
       [0., 0.],
       [1., 0.],
       [0., 1.],
       [1., 1.]])
```

```ipython
>>> mesh.cells
array([[0, 1, 3, 2],
       [4, 5, 7, 6]])
```

The result can be converted back to a meshio-mesh.

```
>>> mesh.as_meshio()
<meshio mesh object>
  Number of points: 8
  Number of cells:
    quad: 2
```

**Comment by keurfonluu** [comment]
It's straightforward with PyVista.

```python
import pyvista as pv

mesh = pv.from_meshio(a) + pv.from_meshio(b)
```

**Reply by adtzlr** [reply]
That's very nice, thanks for that! Also, PyVista merges duplicated points automatically.

**Reply by reox** [reply]
this is really neat! Need to remember that :)
